proposed/2022/PrivateAssetIndepdentFromExcludeAsset.md

```diff
@@ -0,0 +1,137 @@
+# Opt-in `PrivateAssets` asset flow control option precedence over `IncludeAssets/ExcludeAssets` option
+
+- Author Name <https://github.com/erdembayar>
+- Start Date 2022-12-07
+- GitHub Issue <https://github.com/NuGet/Home/issues/6938>
+- GitHub PR <https://github.com/NuGet/NuGet.Client/pull/4976>
+
+## Summary
+
+<!-- One-paragraph description of the proposal. -->
+Currently `IncludeAssets/ExcludeAssets` options take precedence over `PrivateAssets` asset control option when creating package, so in many cases `IncludeAssets/ExcludeAssets` options completely eclipse the `PrivateAssets` option so doesn't let assets flow to consuming parent project. This proposal introduces new a `PackPrivateAssetsFlow` opt-in property so it let `PrivateAssets` asset control option takes precedence over `IncludeAssets/ExcludeAssets` options to enable asset flow to consuming parent projects.
+
+## Motivation
+
+<!-- Why are we doing this? What pain points does this solve? What is the expected outcome? -->
+A package author should be able control which asset flow to parent consuming projects when this project is consumed as package. It enables parent projects to consume transitive packages without having to directly reference them, therefore it reduces the number of packages developers need to keep track of.
+We couldn't make this default experience because it'll break customers who rely on current behavior.
+
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
+The new ` PackPrivateAssetsFlow` property only affects pack operation (more specifically nuspec in nupkg file), but doesn't affect restore experience for current project so there would be no change in `project.assets.json` lock file.
```

Comment:
So, this doesn't impact P2P references for this project? PackageReference in a project tends to propagate to 'parent' projects across P2P as well as those using PackageReference, and PrivateAssets impacts both of them. Is the fix required in both places? In other words, if I have a P2P to this project, and I set PrivateAssets="none" IncludeAssets="all", does the parent project (via P2P) have the same problem that a packed project have? And if so shouldn't we fix that as well?

Reply:
Could you verify if I understood correctly? 
1. Create a parent project `ClassLibrary1` and reference the current project like below.
```  
<ItemGroup>
      <ProjectReference Include="..\refissue\refissue.csproj" PrivateAssets="none" IncludeAssets="compile" />
  </ItemGroup>
```
2. Now pack the `ClassLibrary1` project.
Currently actual `nuspec` file from `ClassLibrary1.1.0.0.nupkg` looks like below:

```
    <dependencies>
      <group targetFramework="net7.0">
        <dependency id="refissue" version="1.0.0-beta" exclude="Runtime,Build,Native,Analyzers,BuildTransitive" />
      </group>
    </dependencies>
```

I believe expected experience would be:

```
    <dependencies>
      <group targetFramework="net7.0">
        <dependency id="refissue" version="1.0.0-beta" include="All" />
      </group>
    </dependencies>
```

Thank you.

Reply:
No, the scenario I'm thinking of doesn't pack `ClassLibrary1.csproj`. It's just a question of whether `ClassLibrary1` itself will get the assets from Microsoft.Windows.CsWin32 through `refissue.csproj` that are prescribed by the `PrivateAssets` attribute in `refissue.csproj`.

IOW, Consider that refissue.csproj has a PackageReference to CsWin32 with `PrivateAssets="none" ExcludeAssets="all"`. Then `ClassLibrary1.csproj` has a simple ProjectReference to `refissue.csproj`. What assets from CsWin32 will ClassLibrary1 get? I hope all of them, which is equivalent to what it would get if `ClassLibrary1` had a *Package*Reference to refissue.

Reply:
It looks like similar issue exists for P2P scenario.

1. Left side is referencing newly generated nupkg with opt-in property. 
```
  <ItemGroup>
    <PackageReference Include="refissue" Version="1.0.0-beta" />
  </ItemGroup>
```

Package ref inside `refissue.csproj`:
```
  <ItemGroup>
    <PackageReference Include="Microsoft.Windows.CsWin32" Version="0.2.138-beta" PrivateAssets="none" ExcludeAssets="all" />
  </ItemGroup>
```
2. Right side is just simple Project reference to `refissue.csproj`
```
  <ItemGroup>
    <ProjectReference Include="..\refissue\refissue.csproj" />
  </ItemGroup>
```

Package ref inside `refissue.csproj`:
```
  <ItemGroup>
    <PackageReference Include="Microsoft.Windows.CsWin32" Version="0.2.138-beta" PrivateAssets="none" ExcludeAssets="all" />
  </ItemGroup>

```
Below is prop file difference.
![image](https://user-images.githubusercontent.com/8766776/206540935-898be4e0-6cb3-4d9c-a77e-84d6b4e60555.png)

Reply:
the proposal for project-to-project should also cover where the opt-in in needs to be set, so cover the scenarios of:

P1 -> P2, and then the full truth table for what happens when each individual project opts in.

Reply:
The name `PackPrivateAssetsFlow` would need some workshopping given the feedback.

Reply:
> The name PackPrivateAssetsFlow would need some workshopping given the feedback.

I renamed to `PrivateAssetIndependent` which I believe more represents what it does now.

Reply:
> the proposal for project-to-project should also cover where the opt-in in needs to be set, so cover the scenarios of:
P1 -> P2, and then the full truth table for what happens when each individual project opts in.

I added table, it's not full truth table. For example `Include` = `true` means Exclude = `false`. Please let me know if I need to extend it.

Reply:
> So, this doesn't impact P2P references for this project?

Keep in mind that Include/Private/Exclude Assets does not work on P2P references: https://github.com/NuGet/Home/issues/6098 (I'm sure there are other bugs too. This type of bug doesn't seem to have been de-duplicated well)

It would be great for that bug to be fixed, but it might be a lot more effort than implementing this feature only for PackageReference.

> I renamed to PrivateAssetIndependent which I believe more represents what it does now

I'm afraid that if I had not read the original feature request, that this name would not make any sense to me. From what I can tell, this feature is "treat PrivateAssets as transitive". However, for this exact reason, I don't like the feature proposal as is.  Normally, PrivateAssets means "use asset in current project, but do not flow transitively".  This feature is making it "do not use asset in current project and do flow transitively". In other words, it's completely inverting the meaning/intent of `PrivateAssets`.

Unless I still don't understand something well, I think we need to consider a different approach, as I think the current proposal will confuse customers.

Reply:
> From what I can tell, this feature is "treat PrivateAssets as transitive"

I don't know what that would even mean. It isn't how I would summarize this. I would summarize the feature as "Respect PrivateAssets". It literally should just make nuget take PrivateAssets metadata at face-value rather than conflate it with include/exclude assets.

> Normally, PrivateAssets means "use asset in current project, but do not flow transitively".

This is incorrect. PrivateAssets says absolutely nothing about the use in the current project. It _only_ determines what assets from this reference bleed into parent projects or as nuget dependencies. 

> This feature is making it "do not use asset in current project and do flow transitively". In other words, it's completely inverting the meaning/intent of PrivateAssets.

Not at all. This new feature doesn't change the meaning of PrivateAssets at all. It just makes it so that if you *also* set Include/Exclude assets that these other metadata do not *also* silently change what flows, which is something PrivateAssets is supposed to control.

Reply:
Added P2P scenario in `summary` and `motivation`, `functional explanation` parts.

Reply:
P2P scenario already included in spec.